### PR TITLE
Handle Foreign Key Record Creation

### PR DESCRIPTION
**Pull request by dan-tang-ssd**
PR to accompany https://github.com/stats4sd/tape-data-system/pull/67

It contains below changes:
1. Adds new config items for submission foreign key handling class and method name.
2. Moves business logic for foreign key data retrieval from submodule to main repo.
3. Keeps package generic without application specific business logic.

### CODE REVIEW

File: src/Services/OdkLinkService.php [hunk 1]
```diff
@@ -5,22 +5,20 @@
...
-use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\DB;
 use Illuminate\Support\Facades\Http;
 use Maatwebsite\Excel\Facades\Excel;
 use Illuminate\Support\Facades\Cache;
 use Illuminate\Support\Facades\Schema;
 use Illuminate\Database\Eloquent\Model;
 use Illuminate\Support\Facades\Storage;
 use App\Models\SurveyData\SimpleFormMain;
+use App\Services\HelperService;
```
Comment (dave-mills) [inline src/Services/OdkLinkService.php]: We shouldn't rely on the main app having this exact class.

File: src/Services/OdkLinkService.php [hunk 2]
```diff
@@ -575,7 +573,7 @@ public function getSubmissions(Xlsform $xlsform): int
             ->json();
 
         // only process new submissions
-        $resultsToAdd = Collect($results['value'])->whereNotIn('__id', $xlsform->submissions->pluck('odk_id')->toArray());
+        $resultsToAdd = Collect($results['value'])->whereNotIn('__id', $xlsform->submissions()->withTrashed('soft_deletes')->pluck('odk_id')->toArray());
```
Comment (dave-mills) [inline src/Services/OdkLinkService.php:573]: Nice 👍

File: src/Services/OdkLinkService.php [hunk 3]
```diff
@@ -726,35 +724,132 @@ private function storeMainSurveyToCustomTable(Xlsform $xlsform, $entry, XlsformT
...
+                // dump('$fullItemPath : ' . $fullItemPath);
+
+                $value = Arr::get($entry, $fullItemPath);
                 // dump($schemaItem['name'] . ' : ' . $value);
+            }
+
+
+            // if app developer has defined a method of creating foreign key record in submission content, call that method:
+            $class = config('filament-odk-link.submission.foreign_key_process_method.class');
+            $method = config('filament-odk-link.submission.foreign_key_process_method.method');
```
Comment (dave-mills) [inline src/Services/OdkLinkService.php:792]: I think this works for now. Probably we should eventually move towards a set of interfaces / traits that give a clearer set of requirements to developers (and with the benefit that IDEs can read interfaces and tell you when you have missed something important.

**Review by dave-mills — CHANGES_REQUESTED** [review]
I think the only change we need here immediately is to make sure the helper method being relied on is in the package, and not in the main app. 

Apart from that, this works well. My main comments are more of a longer-term design approach, which is beyond the scope of this PR. (Things like moving towards interfaces and traits that will make it clear to devs what they should do to make a data Model able to be "updatedFromOdk").

File: src/Services/OdkLinkService.php (hunk 1 above)
Comment (dan-tang-ssd) [reply src/Services/OdkLinkService.php]: Oh... yes.

HelperService will be called from main repo class SubmissionForeignKeyController.php.
I have just removed the class reference to main repo class in OdkLinkService.php

File: src/Services/OdkLinkService.php (hunk 2 above)
Comment (dan-tang-ssd) [reply src/Services/OdkLinkService.php:573]: Er.... this change is not made by me...

File: src/Services/OdkLinkService.php (hunk 3 above)
Comment (dan-tang-ssd) [reply src/Services/OdkLinkService.php:792]: Yes, this would be cleaner and easier for developer to follow.

**Review by dave-mills — APPROVED** [review]

**Pull request merged**